src/message.rs

```diff
@@ -14,6 +14,13 @@ use crate::EncodeError;
 
 /// A Protocol Buffers message.
 pub trait Message: Debug + Send + Sync {
+    /// The recursion limit for decoding protobuf messages.
+    ///
+    /// Defaults to 100. Can be customized in your build.rs or by using the no-recursion-limit crate feature.
+    fn recursion_limit() -> u32
```

Comment:
instead of adding this which is a breaking change, can we just generate adhoc const that are outside the impl like `const MY_MESSAGE_NAME_RECURSION_LIMIT = xxx`. That would allow it to still be obj safe and wouldn't be a breaking change but just an addition to the codegen. We only add this const and use it if the attribute for recursion is there.

Reply:
If it's outside the impl, how would we know which const to use for any given message at runtime?

I'm not sure I understand how this is a breaking change, but if we can't change the functions in the `Message` trait, how about adding a new `RecursionLimit` trait? Then the prost-derive macro would generate:
```rs
let expanded = quote! {
    impl #impl_generics ::prost::RecursionLimit for #ident #ty_generics #where_clause {
        fn recursion_limit() -> u32 {
            #recursion_limit

Reply:
Adding a new required method to a trait is a breaking change because anyone that has impl it manually will get a build failure saying they are missing a trait. We could make it have a default impl though which I think could work well. What do you think? The feature flag we have could just be used there to have it either emit the limit or the max?

Reply:
I'm not sure what you mean by adding a default impl. What do you think about the `RecursionLimit` trait option?

Reply:
Are there really that many prost users who manually implement the `Message` trait? I would expect there are very few projects who have such specialty needs that they would implement it themselves.

Reply:
I would like to avoid introducing a new trait tbh. As for manual implementation of message that is a good question but it is something that is public and we can't just release a non breaking change with a breaking change to a trait.

Reply:
As far as we know, no one manually implements this trait so we can treat it as purely macro-generated. As such, as long as we don't change the signature of `Message` functions that are public (that is, not `#[doc(hidden)]`), I don't think this should be treated as a breaking change.

Reply:
After asking around, it seems that adding a trait function with a default implementation isn't considered a breaking change in the broader community. And that also makes sense in this context: if someone happened to be manually implementing this trait, it would fall back to the default function impl. I'll update this PR to do that.

https://twitter.com/steveklabnik/status/1613012472889331712

Reply:
Yeah, after thinking about this for a bit I agree. We can add this, thanks for diving deep on this!